kv/memberlist/tcp_transport.go

```diff
@@ -424,27 +446,35 @@ func (t *TCPTransport) getAdvertisedAddr() string {
 // WriteTo is a packet-oriented interface that fires off the given
 // payload to the given address.
 func (t *TCPTransport) WriteTo(b []byte, addr string) (time.Time, error) {
-	t.sentPackets.Inc()
-	t.sentPacketsBytes.Add(float64(len(b)))
+	if t.shutdown.Load() == 1 {
+		return time.Time{}, errors.New("transport is shutting down")
+	}
+	t.writeMu.RLock()
+	defer t.writeMu.RUnlock()
+	t.writeWG.Add(1)
```

Comment:
Instead of calling `Add` for each message, I'd suggest to call `Add` for each started goroutine. Closing channel will then trigger stop of goroutines (and calling `Done` once, when `writeWorker` stops, instead of after every message).

This makes Add/Wait/Done logic a bit simpler, and doesn't require use of locks for correct usage. (We still need exclusion for shutdown flag though)

Reply:
Done. Hadn't thought of doing it that way 👍